pkg/ingester/retention.go

```diff
@@ -39,183 +66,363 @@ func defaultRetentionPolicy() retentionPolicy {
-	})
-
-	return blocks, nil
+	return deleted
 }
 
-func (e *retentionPolicyEnforcer) cleanupBlocksWhenHighDiskUtilization(ctx context.Context) error {
-	var volumeStatsPrev *diskutil.VolumeStats
-	volumeStatsCurrent, err := e.volumeChecker.HasHighDiskUtilization(e.dbConfig.DataPath)
+// EnforceHighDiskUtilization will run more aggressive disk cleaning if high
+// disk utilization is detected. It returns true if high disk utilization was
+// detected, along with the number of files deleted and the estimated bytes
+// recovered. If no high disk utilization was detected, false is returned.
+func (dc *diskCleaner) EnforceHighDiskUtilization(ctx context.Context) (int, int, bool) {
+	volumeStats, err := dc.volumeChecker.HasHighDiskUtilization(dc.config.DataPath)
 	if err != nil {
-		return err
+		level.Error(dc.logger).Log(
+			"msg", "failed run high disk cleanup, failed to check disk utilization",
+			"err", err,
+		)
+		return 0, 0, false
 	}
+
 	// Not in high disk utilization, nothing to do.
-	if !volumeStatsCurrent.HighDiskUtilization {
-		return nil
+	if !volumeStats.HighDiskUtilization {
+		return 0, 0, false
 	}
-	// Get all block across all the tenants. Any block
-	// produced or imported during the procedure is ignored.
-	blocks, err := e.localBlocks(e.dbConfig.DataPath)
+
+	tenantIDs, err := dc.blockManager.GetTenantIDs(ctx)
 	if err != nil {
-		return err
+		level.Error(dc.logger).Log(
+			"msg", "failed run high disk cleanup, could not read tenant ids",
+			"err", err,
+		)
+		return 0, 0, true
 	}
 
-	for volumeStatsCurrent.HighDiskUtilization && len(blocks) > 0 && ctx.Err() == nil {
-		// When disk utilization is not lower since the last loop, we end the
-		// cleanup there to avoid deleting all blocks when disk usage reporting
-		// is delayed.
-		if volumeStatsPrev != nil && volumeStatsPrev.BytesAvailable >= volumeStatsCurrent.BytesAvailable {
-			level.Warn(e.logger).Log("msg", "disk utilization is not lowered by deletion of a block, pausing until next cycle")
+	blocks := make([]*tenantBlock, 0)
+	for _, tenantID := range tenantIDs {
+		tenantBlocks, err := dc.blockManager.GetBlocksForTenant(ctx, tenantID)
+		if err != nil {
+			level.Error(dc.logger).Log(
+				"msg", "failed to get blocks for tenant",
+				"tenantID", tenantID,
+				"err", err,
+			)
+
+			// Keep trying to read blocks from other tenants.
+			continue
+		}
+
+		blocks = append(blocks, tenantBlocks...)
+	}
+
+	// Sort by uploaded, then age (oldest first).
+	sort.Sort(blocksByUploadAndAge(blocks))
+
+	prevVolumeStats := &diskutil.VolumeStats{}
+	filesDeleted := 0
+	for _, block := range blocks {
+		// Delete a block.
+		err = dc.blockManager.DeleteBlock(ctx, block)
+		switch {
+		case os.IsNotExist(err):
+			level.Warn(dc.logger).Log(
+				"msg", "failed to delete uploaded block, does not exist",
+				"err", err,
+				"path", block.Path,
+			)
+			return filesDeleted, int(volumeStats.BytesAvailable - prevVolumeStats.BytesAvailable), true
```

Comment:
could we `continue` here instead or is there a reason to prefer to exit on the first error?

Reply:
It's possible we could continue. The original implementation stopped on this error, so I mirrored that behavior.

https://github.com/grafana/pyroscope/blob/30f69c48054273649d316c39502a597f7d54e973/pkg/ingester/retention.go#L182-L185

Reply:
didn't notice that, lets keep it as it was for now then